File: myulib-core/src/main/java/com/myudog/myulib/api/MyuVFXManager.java
package com.myudog.myulib.api;

import com.myudog.myulib.api.core.dsl.EffectBuilder;
import net.minecraft.core.particles.ParticleOptions;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.phys.Vec3;

public final class MyuVFXManager {

    public static final MyuVFXManager INSTANCE = new MyuVFXManager();

    
    private MyuVFXManager() {
    }

    public void spawnSpiral(ServerLevel Level, Vec3 center, ParticleOptions particle) {
        EffectBuilder.spawnEffect(center, builder -> {
            builder.duration(40);
            builder.onTick(tick -> {
                double angle = tick * 0.5;
                double radius = 1.5;
                double x = radius * Math.cos(angle);
                double z = radius * Math.sin(angle);
                double y = tick * 0.1;
                VFXCompat.spawnParticles(Level, particle, center.x + x, center.y + y, center.z + z, 1, 0.0, 0.0, 0.0, 0.0);
            });
        });
    }

    public void spawnShockwave(ServerLevel Level, Vec3 center, ParticleOptions particle) {
        EffectBuilder.spawnEffect(center, builder -> {
            builder.duration(15);
            builder.onTick(tick -> {
                double radius = tick * 0.8;
                double density = Math.max(8.0, radius * 10.0);
                for (Vec3 pos : Shapes.CIRCLE.getOutlinePoints(new Vec3(radius, 0.0, radius), density)) {
                    VFXCompat.spawnParticles(Level, particle, center.x + pos.x, center.y + pos.y, center.z + pos.z, 1, 0.0, 0.1, 0.0, 0.02);
                }
            });
        });
    }
}

File: src/main/kotlin/com/myudog/myulib/api/MyuVFXManager.kt
package com.myudog.myulib.api

import com.myudog.myulib.api.dsl.EffectBuilder
import com.myudog.myulib.api.dsl.spawnEffect
import net.minecraft.particle.ParticleEffect
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.math.Vec3d
import kotlin.math.cos
import kotlin.math.sin

object MyuVFXManager {

    /**
     * 預設特效：螺旋升天 (SpiralUp)
     */
    fun spawnSpiral(world: ServerWorld, center: Vec3d, particle: ParticleEffect) {
        spawnEffect(center) {
            duration(40)
            onTick { tick ->
                val angle = tick * 0.5
                val radius = 1.5
                val x = radius * cos(angle)
                val z = radius * sin(angle)
                val y = tick * 0.1

                // offsets (dx, dy, dz, speed) then count last to match various mappings
                VFXCompat.spawnParticles(world, particle, center.x + x, center.y + y, center.z + z, 1, 0.0, 0.0, 0.0, 0.0)
            }
        }
    }

    /**
     * 預設特效：衝擊波 (Shockwave)
     */
    fun spawnShockwave(world: ServerWorld, center: Vec3d, particle: ParticleEffect) {
        spawnEffect(center) {
            duration(15)
            onTick { tick ->
                val radius = tick * 0.8
                val density = (radius * 10.0).toInt().coerceAtLeast(8).toDouble()
                forCircle(radius, density) { pos ->
                    VFXCompat.spawnParticles(world, particle, pos.x, pos.y, pos.z, 1, 0.0, 0.1, 0.0, 0.02)
                }
            }
        }
    }
}